shields: Add support for M5Stack mbus connector and core2 base shield

MrMarteng opened the pull request:
M5Stack base and core2 modules support an extension connector. This PR adds support for this connector. 

It also includes support for the m5stack_core2 base shield, which features a MPU6886 6-axis motion tracker and a SPM1423
microphone.

Commit:
shields: Add support for M5Stack-Core2 base shield.

M5Stack-Core2 comes with a base shield that is connected to the M5Stack
extention connector. It features a MPU6886 6-axis motion tracker
and a SPM1423 microphone.

Signed-off-by: Martin Kiepfer <m.kiepfer@teleschirm.org>

MrMarteng commented:
@kartben I started to prepare support for the m5stack connector. Maybe you find some time to have a look at it. I am happy for feedback.

MrMarteng commented:
@kartben What's your preferred name for the extension header? Currently it's name `m5stack_header`. I also thought about `mbus_header` but mbus is also used by other brands.

Commit:
gpios: shields: Add definition for M5Stack M-Bus port

This commit adds basic support for m5stacks M-Bus extenions port that is
support my core and core2 module.

Signed-off-by: Martin Kiepfer <m.kiepfer@teleschirm.org>

Commit:
sensors: mpu6050: Add support for variant mpu6886

This commit adds support for mpu6886, which has compatible register
layout.

Signed-off-by: Martin Kiepfer <m.kiepfer@teleschirm.org>

MrMarteng commented:
> @kartben What's your preferred name for the extension header? Currently it's name `m5stack_header`. I also thought about `mbus_header` but mbus is also used by other brands.

I named the header `m5stack,mbus-header`. The shield that comes with m5stack is named `m5stack_core2_base`.

kartben commented on boards/shields/m5stack_core2_ext/doc/img/m5stack_core2_ext.jpg:
```shell
 ~/Downloads > optimizt m5stack_core2_ext.jpg
ℹ Optimizing 1 image (lossy)...
✔ m5stack_core2_ext.jpg 
   102.912 KB → 55.072 KB. Ratio: 46%

ℹ Yay! You saved 47.84 KB (46%)

```

jfischer-no commented on boards/xtensa/m5stack_core2/m5stack_mbus_connectors.dtsi:
```diff
@@ -0,0 +1,49 @@
+	};
+};
+
+m5stack_mbus_i2c0: &i2c0 {};
+m5stack_mbus_i2c1: &i2c1 {};
+m5stack_mbus_uart0: &uart1 {};
+m5stack_mbus_uart1: &uart2 {};
+m5stack_mbus_spi: &spi3 {};
```
Please split this path in two, one for the board changes and second where you introduce a new shield. You actual commit message does not reflect the changes.

MrMarteng replied:
You are right. Fixed with latest push

Commit:
boards: m5stack_core2: Add support for mbus connector

This commit adds basic support for the mbus connection port to
m5stack_core2 board.

Signed-off-by: Martin Kiepfer <m.kiepfer@teleschirm.org>

jfischer-no commented on boards/shields/m5stack_core2_ext/Kconfig.defconfig:
```diff
@@ -0,0 +1,9 @@
+
+if M5STACK_CORE2_EXT
+
+config SENSOR
+	default y
```
Board/shields should not enable subsystems/drivers, see also https://docs.zephyrproject.org/latest/hardware/porting/board_porting.html#general-recommendations

MrMarteng replied:
So it's ok then, if the shield isn't working without enabling sensor support?
Should this dependency be highlighted in the documentation or what's the proper way to handle it?

jfischer-no replied:
It is up to your application to enable/disable something. Shields are more about configuration, not enabling. A shield can have multiple configurations for e.g. display and sd card.

jfischer-no commented on boards/shields/m5stack_core2_ext/m5stack_core2_ext.overlay:
```diff
@@ -0,0 +1,19 @@
+		reg = <0x68>;
+	};
+};
+
+/ {
+
+};
```
Please remove these lines.

MrMarteng replied:
fixed

jfischer-no commented on boards/shields/m5stack_core2_ext/m5stack_core2_ext.overlay:
```diff
@@ -0,0 +1,15 @@
+&m5stack_mbus_i2c0 {
+	status = "okay";
+
+	mbus_mpu6886: mbus_mpu6886@68 {
```
Sorry, I missed it, this must be in the form `<device>_<shield>`, ` mpu6886_m5stack_core2_ext: mpu6886@68 {`

MrMarteng replied:
Sorry for the late response. I was on vacation.
PR has been updated

kartben commented on boards/shields/m5stack_core2_ext/Kconfig.shield:
```diff
@@ -0,0 +1,5 @@
+# Copyright (c) 2023 Martin Kiepfer <mrmarteng@teleschirm.org>
+# SPDX-License-Identifier: Apache-2.0
+
+config M5STACK_CORE2_EXT
```
```suggestion
config SHIELD_M5STACK_CORE2_EXT
```

MrMarteng replied:
fixed

kartben commented on boards/shields/m5stack_core2_ext/doc/index.rst:
```diff
@@ -0,0 +1,53 @@
+For example:
+
+.. zephyr-app-commands::
+   :zephyr-app: samples/drivers/mpu6050
```
```suggestion
   :zephyr-app: samples/sensor/mpu6050
```

kartben reviewed (commented):
I can't seem to be able to run the mpu6050 demo?
```
*** Booting Zephyr OS build v3.5.0-rc1-49-gc060ec527fa1 ***
Device mbus_mpu6886@68 is not ready
```

kartben commented:
> I can't seem to be able to run the mpu6050 demo?
> 
> ```
> *** Booting Zephyr OS build v3.5.0-rc1-49-gc060ec527fa1 ***
> Device mbus_mpu6886@68 is not ready
> ```

I am suspecting issues with the AXP192, I see different behaviors /errors with cold boot vs. soft reset

kartben commented:
Also, MPU6050 seemed pretty broken? See #63446

MrMarteng commented:
> > I can't seem to be able to run the mpu6050 demo?
> > ```
> > *** Booting Zephyr OS build v3.5.0-rc1-49-gc060ec527fa1 ***
> > Device mbus_mpu6886@68 is not ready
> > ```
> 
> I am suspecting issues with the AXP192, I see different behaviors /errors with cold boot vs. soft reset

I cannot reproduce it yet
```
uart:~$ *** Booting Zephyr OS build zephyr-v3.4.0-4871-g9a969754e08d ***
Hello World! m5stack_core2
device list
devices:
- rtc@3ff48000 (READY)
- gpio@3ff44800 (READY)
- gpio@3ff44000 (READY)
- trng@3ff75144 (READY)
- uart@3ff40000 (READY)
- i2c@3ff53000 (READY)
- axp192@34 (READY)
- LDO3 (READY)
- LDO2 (READY)
- DCDC3 (READY)
- DCDC1 (READY)
- axp192_gpio (READY)
- ft5336@38 (READY)
- mbus_mpu6886@68 (READY)
uart:~$ sensor get mbus_mpu6886@68
channel idx=3 accel_xyz shift=4 value=25589900000ns, (0.148439, -0.272939, 10.606312)
channel idx=7 gyro_xyz shift=4 value=25589900000ns, (-0.120307, -0.263131, 0.298436)
uart:~$ sensor get mbus_mpu6886@68
channel idx=3 accel_xyz shift=4 value=27917600000ns, (0.047883, 8.683768, 3.193864)
channel idx=7 gyro_xyz shift=4 value=27917600000ns, (0.363853, -0.402757, 0.528260)
uart:~$ sensor get mbus_mpu6886@68
channel idx=3 accel_xyz shift=4 value=29257400000ns, (9.732428, -0.773328, 1.125273)
channel idx=7 gyro_xyz shift=4 value=29257400000ns, (-0.208640, -0.151350, 0.263796)
```
It doesn't make a difference on soft/hard reboot

Can you please check the priorities of `CONFIG_REGULATOR_AXP192_INIT_PRIORITY`, `CONFIG_I2C_INIT_PRIORITY` and `CONFIG_SENSOR_INIT_PRIORITY`. My default configuration is:

```
#define CONFIG_REGULATOR_AXP192_INIT_PRIORITY 76
#define CONFIG_GPIO_AXP192_INIT_PRIORITY 80
#define CONFIG_GPIO_HOGS_INIT_PRIORITY 81
#define CONFIG_I2C_INIT_PRIORITY 50
#define CONFIG_SENSOR_INIT_PRIORITY 90
```

Which project did you use for testing?

kartben commented:
> Which project did you use for testing?

It might have been my fault, sorry. I was testing against #63375, looks like it requires a slightly larger task than 2KB when running on the Core2. Works now :)

MrMarteng commented:
> > Which project did you use for testing?
> 
> It might have been my fault, sorry. I was testing against #63375, looks like it requires a slightly larger task than 2KB when running on the Core2. Works now :)

Yes, I noticed this as well and increased the default stack size for m5stack in #62507 :-/

kartben commented on boards/shields/m5stack_core2_ext/doc/index.rst:
```diff
@@ -0,0 +1,53 @@
+********
+
+M5Stack-Core2 comes with a base shield that is connected to the M5Stack
+extension connector. It features a MPU6886 6-axis motion tracker and a SPM1423
```
nit:
```suggestion
extension connector. It features an MPU6886 6-axis motion tracker and a SPM1423
```

MrMarteng replied:
fixed

kartben commented on boards/shields/m5stack_core2_ext/doc/index.rst:
```diff
@@ -0,0 +1,53 @@
+   :shield: m5stack_core2_ext
+   :goals: build
+
+References
+**********
+
+.. target-notes::
+
+.. _datasheet:
+   https://docs.m5stack.com/en/core/core2
```
drop unless you actually have links/footnotes.

MrMarteng replied:
fixed

kartben reviewed (changes requested):
Neat! See my comments as well as earlier comments from today.

kartben commented on boards/xtensa/m5stack_core2/doc/index.rst:
```diff
@@ -74,6 +74,7 @@ of the M5Stack Core2 board.
 +------------------+--------------------------------------------------------------------------+
 || 3-axis          || The `MPU-6886`_ is a 6-axis MotionTracking device that combines a       |
 || gyrosopce       || 3-axis gyroscope and a 3-axis accelerometer.                            |
```
well... now that you mention it ... :D

```suggestion
|| gyroscope       || 3-axis gyroscope and a 3-axis accelerometer.                            |
```

MrMarteng replied:
I am not an expert on these sensors.I reworked the used terms in both documents (m5stack_core2 and m5stack_core2_ext).

MrMarteng commented:
further comments?

carlescufi commented:
@erwango ping
